environment_secrets.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2017 Rohith All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package marathon
+
+import (
+	"encoding/json"
+	"fmt"
+)
+
+// Alias aliases the Application struct so that it will be marshaled/unmarshaled automatically
+type Alias Application
+
+// TmpEnvSecret holds the secret values deserialized from the environment variables field
+type TmpEnvSecret struct {
+	Secret string `json:"secret,omitempty"`
+}
+
+// TmpSecret holds the deserialized secrets field in a Marathon application configuration
+type TmpSecret struct {
+	Source string `json:"source,omitempty"`
+}
+
+// UnmarshalJSON unmarshals the given Application JSON into an environment variables and secrets.
+func (app *Application) UnmarshalJSON(b []byte) error {
+	aux := &struct {
+		*Alias
+		Env     map[string]interface{} `json:"env,omitempty"`
+		Secrets map[string]TmpSecret   `json:"secrets,omitempty"`
+	}{
+		Alias: (*Alias)(app),
+	}
+	if err := json.Unmarshal(b, aux); err != nil {
+		return err
+	}
+	env := make(map[string]string)
+	secrets := make(map[string]Secret)
+
+	if aux.Env != nil {
+		for k, v := range aux.Env {
+			if s, ok := v.(string); ok {
+				env[k] = s
+				continue
+			}
+			tmp, err := json.Marshal(v)
```

Comment:
The secrets in env deserialize as a map[string]string, so instead of messing with relfection and all the error handling that requires.  I marshal it, then unmarshal it into a TmpEnvSecret.

To be honest, I don't know exactly how to use reflection to do this with possible nil values, yet, but if it makes more sense to do it that way I can work on it.